JVM_IR: defer some branch optimizations to codegen.

ting-yuan opened the pull request:
Specifically, defer the removal of hand-written "if (true)" from
JvmBuiltinOptimizationLowering into codegen so that appropriate debug
info (and a NOP) can be inserted.

Some optimization in visitBlock in JvmBuiltinOptimizationLowering depends on the removal of the "if (true)" in visitWhen. Therefore, it cannot be deferred (unless we move the logic into visiBlock to make it even more complicated). Fortunately, the optimization in visitBlock only cares about compiler generated constructs, which is transparent to users and need no debug info. To defer the optimization for non-compiler generated constructs, cl:1260 marks "hand-written" ifs by setting origin to IF so that
* compiler generate stuffs have origin == null and can be removed in IR.
* hand-written stuffs have origin == IF and can only be removed, with correct debug info generated, in ExpressionCodegen.

ting-yuan commented:
In the first commit, only BranchingExpressionGenerator.kt and ExpressionHelpers.kt contain "real" changes. The remainings are just corresponding changes in the expected outputs.

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt:
```diff
@@ -695,9 +696,16 @@ class ExpressionCodegen(
-        if (!elseBranch) {
+        // Else branch comes from "if (ALWAYS_TRUE)", where ALWAYS_TRUE can be handwritten, inferred by compiler,
+        // or a construct of IrElseBranch.
+        val isElseBranch = branch.condition.isTrueConst()
```
@ting-yuan Please name variable properly: 'true' condition could also be in non-else

max-kammerer commented on compiler/testData/codegen/bytecodeText/lineNumbers/ifElseConstTrue.kt:
```diff
@@ -0,0 +1,8 @@
+fun foo() {
+    if (true) {
+        System.out?.println()
```
Don't use print in tests

max-kammerer commented on compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/BranchingExpressionGenerator.kt:
```diff
@@ -77,7 +77,7 @@ class BranchingExpressionGenerator(statementGenerator: StatementGenerator) : Sta
         if (irBranches.size == 1) {
```
@dnpetrov Shoudn't we use IF origin  for irBranches.size > 1 below:
```
val irWhen = IrWhenImpl(ktIf.startOffsetSkippingComments, ktIf.endOffset, resultType, IrStatementOrigin.WHEN)
```

@ting-yuan Please also add next test:
```
fun cond() = false

fun bar() {}

fun foo() {
    if (cond()) {
        bar()
    } else if (true) {
        bar()
    } else {
        bar()
    }
}
// 1 LINENUMBER 6
// 1 LINENUMBER 8
```

max-kammerer replied:
Similiar tests with 'false' would be also appreciated

ting-yuan commented:
Hi @max-kammerer , please take a look again. Changes are:
1. handled `if (false)` as well and added test cases.
2. refactored `genIfWithBranches` a little bit.
3. renamed the tests for true.
4. removed `println` from the tests. In fact, I copied them from existing tests: `ifElse.kt` and `singleThen.kt`. I'm not sure why they use `println` so I didn't touch them. Let me know if you'd like them be replaced, too.

max-kammerer commented:
@ting-yuan Many Thanks!
